v2/stream/new/client.go

```diff
@@ -0,0 +1,336 @@
+}
+
+var connCreator = func(ctx context.Context, u url.URL) (conn, error) {
+	return newNhooyrWebsocketConn(ctx, u)
+}
+
+func constructURL(host, feed string) (url.URL, error) {
+	scheme := "wss"
+	ub, err := url.Parse(host)
+	if err != nil {
+		return url.URL{}, err
+	}
+	switch ub.Scheme {
+	case "http", "ws":
+		scheme = "ws"
+	}
+
+	return url.URL{Scheme: scheme, Host: ub.Host, Path: "/v2/" + feed}, nil
+}
+
+var ErrConnectCalledMultipleTimes = errors.New("Connect called multiple times")
+
+// Connect establishes a connection for c using its configuration and reestablishes it when errors occur.
+// It blocks until the connection has been established for the first time (or it failed to do so).
+// Should only be called once!
+//
+// It returns after the connection was established for the first time or if it couldn't be established
+// after retrying.
+func (c *client) Connect(ctx context.Context) error {
+	err := ErrConnectCalledMultipleTimes
+	c.connectOnce.Do(func() {
+		err = c.connect(ctx)
+		if err != nil {
+			c.connErrChan <- err
+		}
+		c.connectCalled = true
+	})
+	return err
+}
+
+func (c *client) connect(ctx context.Context) error {
+	u, err := constructURL(c.host, c.feed)
+	if err != nil {
+		return err
+	}
+
+	successCh := make(chan struct{})
+	go c.maintainConnection(ctx, u, successCh)
+
+	select {
+	case <-successCh:
+		return nil
+	case err := <-c.connErrChan:
+		return err
+	}
+}
+
+func (c *client) Error() <-chan error {
+	return c.connErrChan
+}
+
+// maintainConnection initializes a connection to u, starts the necessary goroutines
+// and recreates them if there was an error as long as reconnectLimit consecutive
+// connection initialization errors don't occur. It closes successCh upon the first
+// successful connection intialization
+func (c *client) maintainConnection(ctx context.Context, u url.URL, successCh chan<- struct{}) {
+	var connError error
+	failedAttemptsInARow := 0
+	connectedAtLeastOnce := false
+
+	defer func() {
+		// If there is a pending sub change we should terminate that
+		c.pendingSubChangeMutex.Lock()
+		defer c.pendingSubChangeMutex.Unlock()
+		if c.pendingSubChange != nil {
+			c.pendingSubChange.result <- ErrSubChangeInterrupted
+		}
+		c.pendingSubChange = nil
+		c.hasTerminated = true
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if !connectedAtLeastOnce {
+				c.logger.Warnf("datav2stream: cancelled before connection could be established, last error: %v", connError)
+				err := fmt.Errorf("cancelled before connection could be established, last error: %w", connError)
+				c.connErrChan <- err
+			}
+			return
+		default:
+			if c.reconnectLimit != 0 && failedAttemptsInARow >= c.reconnectLimit {
+				c.logger.Errorf("datav2stream: max reconnect limit has been reached, last error: %v", connError)
+				c.connErrChan <- fmt.Errorf("max reconnect limit has been reached, last error: %w", connError)
+				return
+			}
+			time.Sleep(time.Duration(failedAttemptsInARow) * c.reconnectDelay)
```

Comment:
minor: I know that this works but a little bit strange to me. `time.Duration(4)` means `4ns` and we multiply it with `c.reconnectDelay`.

Reply:
I'm not aware of any other way due to how types work in the language.